x-pack/plugins/cloud_security_posture/server/routes/stats/stats.ts

```diff
@@ -0,0 +1,220 @@
+    totalPassed,
+    totalFailed,
+  };
+
+  assertBenchmarkStats(stats);
+
+  return stats;
+};
+
+export const getBenchmarksStats = async (
+  esClient: ElasticsearchClient,
+  cycleId: string,
+  benchmarks: string[]
+): Promise<BenchmarkStats[]> => {
+  const benchmarkPromises = benchmarks.map((benchmark) => {
+    const benchmarkFindings = esClient.count(getFindingsEsQuery(cycleId, undefined, benchmark));
+    const benchmarkPassedFindings = esClient.count(
+      getFindingsEsQuery(cycleId, RULE_PASSED, benchmark)
+    );
+    const benchmarkFailedFindings = esClient.count(
+      getFindingsEsQuery(cycleId, RULE_FAILED, benchmark)
+    );
+
+    return Promise.all([benchmarkFindings, benchmarkPassedFindings, benchmarkFailedFindings]).then(
+      ([benchmarkFindingsResult, benchmarkPassedFindingsResult, benchmarkFailedFindingsResult]) => {
+        const totalFindings = benchmarkFindingsResult.body.count;
+        const totalPassed = benchmarkPassedFindingsResult.body.count;
+        const totalFailed = benchmarkFailedFindingsResult.body.count;
+        const postureScore = calculatePostureScore(totalFindings, totalPassed, totalFailed);
+        const stats = {
+          name: benchmark,
+          postureScore,
+          totalFindings,
+          totalPassed,
+          totalFailed,
+        };
+
+        assertBenchmarkStats(stats);
+        return stats;
+      }
+    );
+  });
+
+  return Promise.all(benchmarkPromises);
+};
+
+export const getResourceTypesAggs = async (
+  esClient: ElasticsearchClient,
+  cycleId: string
+): Promise<CloudPostureStats['resourceTypesAggs']> => {
+  const resourceTypesQueryResult = await esClient.search<unknown, ResourceTypeBucket>(
+    getRisksEsQuery(cycleId)
+  );
+
+  const resourceTypesAggs = resourceTypesQueryResult.body.aggregations?.resource_types.buckets;
+  if (!Array.isArray(resourceTypesAggs)) throw new Error('missing resources types buckets');
+
+  return resourceTypesAggs.map((bucket) => {
+    const evalBuckets = bucket.bucket_evaluation.buckets;
+    if (!Array.isArray(evalBuckets)) throw new Error('missing resources types evaluations buckets');
+
+    const failedBucket = evalBuckets.find((evalBucket) => evalBucket.key === RULE_FAILED);
+    const passedBucket = evalBuckets.find((evalBucket) => evalBucket.key === RULE_PASSED);
+
+    return {
+      resourceType: bucket.key,
+      totalFindings: bucket.doc_count,
+      totalFailed: failedBucket?.doc_count || 0,
+      totalPassed: passedBucket?.doc_count || 0,
+    };
+  });
+};
+
+export const defineGetStatsRoute = (router: IRouter, logger: Logger): void =>
+  router.get(
+    {
+      path: STATS_ROUTE_PATH,
+      validate: false,
+    },
+    async (context, _, response) => {
+      try {
+        const esClient = context.core.elasticsearch.client.asCurrentUser;
+        const [benchmarks, latestCycleID] = await Promise.all([
+          getBenchmarks(esClient),
+          getLatestCycleId(esClient),
+        ]);
```

Comment:
You could consider using the [Point in Time ES feature](https://www.elastic.co/guide/en/elasticsearch/reference/current/point-in-time-api.html). It would help ensure that all of the queries done as part of this API call return consistent data.

This is a new feature, Elastic Security is working on adopting it now. [I found this article very helpful.](https://www.elastic.co/blog/get-a-consistent-view-of-your-data-over-time-with-the-elasticsearch-point-in-time-reader)

Reply:
This seems like the correct way to do this. For now I added a TODO to the code, we will take care of this in a future iteration